### PR TITLE
Fyle-cqf2zp Matching Expenses and Create Expense from Transactions in Personal Cards

### PR DESCRIPTION
**Pull request by iravisankar**
localhost_8100_enterprise_personal_cards_matched_expenses(iPhone X) (1)
localhost_8100_enterprise_personal_cards_matched_expenses(iPhone X) (2)

### CODE REVIEW

File: src/app/core/services/personal-cards.service.ts [hunk 1]
```diff
@@ -41,4 +48,34 @@ export class PersonalCardsService {
...
+
+  getMatchedExpenses(amount: number, txnDate: string): Observable<any[]> {
+    return this.apiService
+      .get('/expense_suggestions/personal_cards', {
+        params: {
+          amount,
+          txn_dt: txnDate,
+        },
+      })
+      .pipe(shareReplay(1));
```
Comment (Julias0) [inline src/app/core/services/personal-cards.service.ts:60]: remove shareReplay from the service. This should only be added at a component level
If you want to add a cache use the decorator wala approach

File: src/app/core/services/personal-cards.service.ts [hunk 2]
```diff
@@ -41,4 +48,34 @@ export class PersonalCardsService {
       })
       .pipe(map((res) => res.count));
   }
+
+  getMatchedExpenses(amount: number, txnDate: string): Observable<any[]> {
```
[the remark this answers is not in the document]
Comment (Julias0) [reply src/app/core/services/personal-cards.service.ts:52]: Dont use `any`. Define the model in an interface

File: src/app/core/services/personal-cards.service.ts [hunk 3]
```diff
@@ -41,4 +48,34 @@ export class PersonalCardsService {
...
+
+  getExpenseDetails(transactionSplitGroupId: string): Observable<any> {
+    return this.apiv2Service.get('/expenses', {
+      params: {
+        tx_split_group_id: `eq.${transactionSplitGroupId}`,
+      },
+    });
+  }
+
+  matchExpense(transactionSplitGroupId: string, externalExpenseId: string): Observable<any> {
```
Comment (Julias0) [inline src/app/core/services/personal-cards.service.ts:75]: dont use any

File: src/app/fyle/add-edit-expense/add-edit-expense.page.ts [hunk 4]
```diff
@@ -2880,13 +2925,17 @@ export class AddEditExpensePage implements OnInit {
...
           if (that.mode === 'add') {
-            that.addExpense('SAVE_EXPENSE').subscribe((res: any) => {
-              if (that.fg.controls.add_to_new_report.value && res && res.transaction) {
-                this.addToNewReport(res.transaction.id);
-              } else {
-                that.goBack();
-              }
-            });
+            if (that.isCreatedFromPersonalCard) {
+              that.saveAndMatchWithPersonalCardTxn();
```
Comment (Julias0) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.ts:2929]: is the button loader handled?

File: src/app/fyle/add-edit-expense/add-edit-expense.page.ts [hunk 5]
```diff
@@ -3956,4 +4005,23 @@ export class AddEditExpensePage implements OnInit {
...
   }
+
+  saveAndMatchWithPersonalCardTxn() {
+    const personalCardTxn =
+      this.activatedRoute.snapshot.params.personalCardTxn &&
+      JSON.parse(this.activatedRoute.snapshot.params.personalCardTxn);
+    const externalExpenseId = personalCardTxn.btxn_id;
+    const txn$ = this.etxn$.pipe(switchMap((etxn) => this.transactionService.upsert(etxn.tx)));
+    this.saveExpenseLoader = true;
+    txn$
```
[the remark this answers is not in the document]
Comment (Julias0) [reply src/app/fyle/add-edit-expense/add-edit-expense.page.ts:4016]: ` this.saveExpenseLoader = false;` this should be inside finally

File: src/app/shared/components/route-selector/route-selector-modal/route-selector-modal.component.html [hunk 6]
```diff
@@ -85,8 +85,11 @@
       </ng-container>
     </div>
 
-    <app-fy-alert-info *ngIf="mileageLocations.controls.length === 10" [message]="'You can add a maximum of 10 locations'"
-      [type]="'information'">
+    <app-fy-alert-info
+      *ngIf="mileageLocations.controls.length === 10"
```
Comment (Julias0) [inline src/app/shared/components/route-selector/route-selector-modal/route-selector-modal.component.html:89]: mileageLocations.controls.length >= 10

**Review by pradeep8093 — APPROVED** [review]

File: src/app/fyle/personal-cards-matched-expenses/personal-cards-matched-expenses.page.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    private router: Router,
+    private modalController: ModalController
+  ) {
+    this.txnDetails = this.router.getCurrentNavigation().extras.state.txnDetails;
+  }
+
+  ngOnInit() {}
+
+  ionViewWillEnter() {
+    const txnDate = new Date(this.txnDetails.btxn_transaction_dt).toISOString().slice(0, 10);
```
Comment (Julias0) [inline src/app/fyle/personal-cards-matched-expenses/personal-cards-matched-expenses.page.ts:33]: no no no. We should use moment for formatting the dates. String manipulation on ISO dates is very hacky

**Pull request merged**